# Netflix Exploratory Data Analysis (EDA) using NumPy and Pandas

**Author:** Altamash

## Project Objective

This project performs complete Exploratory Data Analysis (EDA) on the Netflix  
Movies and TV Shows dataset using NumPy, Pandas, and Matplotlib.
### Goals

- Understand dataset structure.
- Clean missing and inconsistent data.
- Perform statistical analysis using NumPy.
- Analyze Netflix content using Pandas.
- Create business insights and visualizations.

In [7]:
# ==========================================
# Import Required Libraries
# ==========================================

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# Notebook display settings
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 20)

print("Libraries imported successfully.")

Libraries imported successfully.


## Load Netflix Dataset

The dataset is loaded from the `data` folder into a Pandas DataFrame.

In [9]:
# ==========================================
# Load Dataset
# ==========================================

df = pd.read_csv("../data/netflix_titles.csv")

print("Dataset loaded successfully.")

Dataset loaded successfully.


In [10]:
# Dataset Shape

df.shape

(8807, 12)

In [12]:
#Display the first five records.
df.head()

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,NaN,United States,"September 25, 2021",2020,PG-13,90 min,Documentaries,"As her father nears the end of his life, filmm..."
1,s2,TV Show,Blood & Water,NaN,"Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...",South Africa,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, TV Dramas, TV Mysteries","After crossing paths at a party, a Cape Town t..."
2,s3,TV Show,Ganglands,Julien Leclercq,"Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...",NaN,"September 24, 2021",2021,TV-MA,1 Season,"Crime TV Shows, International TV Shows, TV Act...",To protect his family from a powerful drug lor...
3,s4,TV Show,Jailbirds New Orleans,NaN,NaN,NaN,"September 24, 2021",2021,TV-MA,1 Season,"Docuseries, Reality TV","Feuds, flirtations and toilet talk go down amo..."
4,s5,TV Show,Kota Factory,NaN,"Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...",India,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, Romantic TV Shows, TV ...",In a city of coaching centers known to train I...


In [13]:
df.tail()

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
8802,s8803,Movie,Zodiac,David Fincher,"Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...",United States,"November 20, 2019",2007,R,158 min,"Cult Movies, Dramas, Thrillers","A political cartoonist, a crime reporter and a..."
8803,s8804,TV Show,Zombie Dumb,NaN,NaN,NaN,"July 1, 2019",2018,TV-Y7,2 Seasons,"Kids' TV, Korean TV Shows, TV Comedies","While living alone in a spooky town, a young g..."
8804,s8805,Movie,Zombieland,Ruben Fleischer,"Jesse Eisenberg, Woody Harrelson, Emma Stone, ...",United States,"November 1, 2019",2009,R,88 min,"Comedies, Horror Movies",Looking to survive in a world taken over by zo...
8805,s8806,Movie,Zoom,Peter Hewitt,"Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...",United States,"January 11, 2020",2006,PG,88 min,"Children & Family Movies, Comedies","Dragged from civilian life, a former superhero..."
8806,s8807,Movie,Zubaan,Mozez Singh,"Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...",India,"March 2, 2019",2015,TV-14,111 min,"Dramas, International Movies, Music & Musicals",A scrappy but poor boy worms his way into a ty...


In [14]:
#List all feature names available in the dataset.

df.columns

Index(['show_id', 'type', 'title', 'director', 'cast', 'country', 'date_added',
       'release_year', 'rating', 'duration', 'listed_in', 'description'],
      dtype='object')

In [21]:
#Count rows and columns separately.

print("numbers of rows : ",df.shape[0])
print("numbers of columns : ",df.shape[1])

numbers of rows :  8807
numbers of columns :  12


## Dataset Information

Understand data types, missing values, and memory usage.

In [23]:

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


## Statistical Summary

Generate summary statistics for numerical columns.

In [24]:
df.describe()

,release_year
count,8807.000000
mean,2014.180198
std,8.819312
min,1925.000000
25%,2013.000000
50%,2017.000000
75%,2019.000000
max,2021.000000


## Categorical Summary

Explore object-type columns.

In [25]:
df.describe(include="object")

,show_id,type,title,director,cast,country,date_added,rating,duration,listed_in,description
count,8807,8807,8807,6173,7982,7976,8797,8803,8804,8807,8807
unique,8807,2,8807,4528,7692,748,1767,17,220,514,8775
top,s1,Movie,Dick Johnson Is Dead,Rajiv Chilaka,David Attenborough,United States,"January 1, 2020",TV-MA,1 Season,"Dramas, International Movies","Paranormal activity at a lush, abandoned prope..."
freq,1,6131,1,19,19,2818,109,3207,1793,362,4


## Data Types

Check data type of each feature.

In [26]:
df.dtypes

show_id         object
type            object
title           object
director        object
cast            object
country         object
date_added      object
release_year     int64
rating          object
duration        object
listed_in       object
description     object
dtype: object

### Observation

date_added should be converted from object to datetime during data cleaning.

## Memory Usage

Check memory consumed by each column.

In [29]:
memory_usage = df.memory_usage(deep=True)

memory_usage

Index               132
show_id          474471
type             480930
title            593151
director         488185
cast            1422387
country          517757
date_added       561028
release_year      70456
rating           470514
duration         493486
listed_in        725748
description     1873527
dtype: int64

In [30]:
total_memory = memory_usage.sum() / (1024**2)

print(f"Total Memory Usage: {total_memory:.2f} MB")

Total Memory Usage: 7.79 MB


## Frequency Analysis

Count Movies and TV Shows.

In [31]:
df["type"].value_counts()

type
Movie      6131
TV Show    2676
Name: count, dtype: int64

In [32]:
df["type"].value_counts(normalize=True) * 100

type
Movie      69.615079
TV Show    30.384921
Name: proportion, dtype: float64

## Unique Value Analysis

In [34]:
df["country"].nunique()

748

In [36]:
df["country"].unique()[:10]

array(['United States', 'South Africa', nan, 'India',
       'United States, Ghana, Burkina Faso, United Kingdom, Germany, Ethiopia',
       'United Kingdom', 'Germany, Czech Republic', 'Mexico', 'Turkey',
       'Australia'], dtype=object)

## EDA Summary Table

Create a compact summary containing data type, missing values, and unique values.

In [37]:
summary = pd.DataFrame({
    "Data Type": df.dtypes,
    "Missing Values": df.isnull().sum(),
    "Unique Values": df.nunique()
})

summary

,Data Type,Missing Values,Unique Values
show_id,object,0,8807
type,object,0,2
title,object,0,8807
director,object,2634,4528
cast,object,825,7692
country,object,831,748
date_added,object,10,1767
release_year,int64,0,74
rating,object,4,17
duration,object,3,220


## Initial Findings

### Key Observations

- The dataset contains 8,807 Netflix titles.
- There are 12 features describing each title.
- type contains two categories: Movies and TV Shows.
- date_added is stored as an object and requires datetime conversion.
- director, cast, and country contain missing values.
- Both numerical and categorical columns are present, requiring different analysis approaches.